src/modules/schnorrsig/main_impl.h

```diff
@@ -264,4 +264,291 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
            secp256k1_fe_equal_var(&rx, &r.x);
 }
 
+
+
+
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("HalfAgg/randomizer")||SHA256("HalfAgg/randomizer"). */
+void secp256k1_schnorrsig_sha256_tagged_aggregation(secp256k1_sha256 *sha) {
+   secp256k1_sha256_initialize(sha);
+    sha->s[0] = 0xd11f5532ul;
+    sha->s[1] = 0xfa57f70ful;
+    sha->s[2] = 0x5db0d728ul;
+    sha->s[3] = 0xf806ffe1ul;
+    sha->s[4] = 0x1d4db069ul;
+    sha->s[5] = 0xb4d587e1ul;
+    sha->s[6] = 0x50451c2aul;
+    sha->s[7] = 0x10fb63e9ul;
+
+    sha->bytes = 64;
+}
+
+/**
+ * Internal helper function for half-aggregation: 
+ * Computes the coefficients zi for i=starting_from to i=n_sigs-1
+ * used for aggregation for a given sequence of messages, keys and and r's.
+ * zs should have space for n_sigs-starting_from many scalars.
+ * Note: We pass pointers to the rs to avoid having to copy them twice
+*/
+static int compute_aggregation_coefficients(
+    secp256k1_scalar* zs,
+    const unsigned char* pubkeys_ser, 
+    const unsigned char** rs,
+    const unsigned char* msgs32, 
+    size_t starting_from,
+    size_t n_sigs) {
+
+
+    uint32_t i;
+    secp256k1_sha256 hash;
+
+    if (n_sigs < starting_from) {
+        return 0;
+    }
+
+    /* Write common prefix pre into tagged hash  */
+    /* Let n' = starting_from */
+    /* pre = (r_0 || pk_0 || mes_0 || .... ||  r_{n'-1} || pk_{n'-1} || mes_{n'-1}) */
+    secp256k1_schnorrsig_sha256_tagged_aggregation(&hash);
+    for (i = 0; i < starting_from; ++i) {
+        /* write r_i */
+        secp256k1_sha256_write(&hash, rs[i], 32);
+        /* write pk_i */
+        secp256k1_sha256_write(&hash, &pubkeys_ser[i*32], 32);
+        /* write mes_i*/
+        secp256k1_sha256_write(&hash, &msgs32[i*32], 32);
+    }
+
+    /* zi = Hash( pre || r_{n'} || pk_{n'} || mes_{n'} || ... || r_{i} || pk_{i} || mes_{i} ) */
+    for (i = starting_from; i < n_sigs; ++i) {
+        unsigned char hashoutput[32];
+        secp256k1_sha256 hashcopy; 
+        /* Write into hash r_i, pk_i, mes_i, r_i */
+        secp256k1_sha256_write(&hash, rs[i], 32);
+        secp256k1_sha256_write(&hash, &pubkeys_ser[i*32], 32);
+        secp256k1_sha256_write(&hash, &msgs32[i*32], 32);
+        /* Copy the hash */
+        memcpy(&hashcopy,&hash,sizeof(hash));
```

Comment:
```suggestion
        hashcopy = hash;
```

structs can be assigned in C ;)